### PR TITLE
Fix: arguments supplied to restore are split with whitespace

### PR DESCRIPTION
**Pull request by YizhePKU**
This PR fixes a bug that causes arguments with whitespaces to be split into multiple arguments during restore.

It contains two commits. The first commit is refactor only. It moves all side effects from `CracRestoreParameters` to the call sites, and changes the type of `CracRestoreParameters::args` from a single string to a `GrowableArray` of strings. Serialization code is also cleaned up a bit.

The second commit fixes the bug by introducing a new pseudo property, `-DCRaCJavaMainArgs`, that is set in `JavaMain`. An instance of `JavaMainArgs` (containing `argc` and `argv`) is stored as extra info of the property, which is later extracted in `Arguments::parse_options_for_restore` and passed to `crac::restore`.

Potential issues:

* We use `putenv` to modify environment variables, which expects `char*`. In this PR, I'm `const_cast`ing from `const char*` to `char*`, since I believe `putenv` doesn't actually modify the string. Is that OK? Maybe rewriting with `setenv` would be better?
* `read_growable_array` is implemented by reading byte-at-a-time from the shared memory. Will it be too slow? I could rewrite it to read everything at once (like the original code did), but the current implementation is cute and I kinda want to keep it.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/crac.git pull/101/head:pull/101` \
`$ git checkout pull/101`

Update a local copy of the PR: \
`$ git checkout pull/101` \
`$ git pull https://git.openjdk.org/crac.git pull/101/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 101`

View PR using the GUI difftool: \
`$ git pr show -t 101`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/crac/pull/101.diff">https://git.openjdk.org/crac/pull/101.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/crac/pull/101#issuecomment-1697287254)

### CODE REVIEW

File: src/hotspot/share/runtime/crac.cpp [hunk 1]
```diff
@@ -446,7 +509,127 @@ Handle crac::checkpoint(jarray fd_arr, jobjectArray obj_arr, bool dry_run, jlong
...
+      fprintf(stderr, "read_jlong truncated");
+    }
+    return false;
+  }
+  return true;
+}
+
+// Write a GrowableArray to fd.
+// On error, return false.
+static bool write_growable_array(int fd, GrowableArray<const char *>* array) {
```
Comment (rvansa) [inline src/hotspot/share/runtime/crac.cpp:551]: Nitpick: can the array be const?

**Review by rvansa — COMMENTED** [review]
I like the more structured serialization/deserialization of the args, however the binary compatibility and reuse of structures makes me worried.

Could you also add a test that demonstrates arguments with spaces, and zero-length arguments?

File: src/hotspot/share/runtime/arguments.hpp [hunk 2]
```diff
@@ -148,6 +148,15 @@ class SystemProperty : public PathString {
 // Helper class for controlling the lifetime of JavaVMInitArgs objects.
 class ScopedVMInitArgs;
 
+// Arguments passed from JavaMain via the property CRaCJavaMainArgs
```
[the remark this answers is not in the document]
Comment (YizhePKU) [reply src/hotspot/share/runtime/arguments.hpp:151]: > Duplicating structure definition and relying on the same ABI is too fragile in my opinion.

If you mean `JavaMainArgs` could change its content without us noticing, then maybe it's better that we define our own struct. Otherwise, duplicating structure definition is actually the same thing as defining it in a header file and then including the header twice, since `#include` in C++ is literally copy-and-paste.

File: src/java.base/share/classes/jdk/crac/Core.java [hunk 3]
```diff
@@ -228,30 +228,27 @@ public String run() {
...
-                } catch (PrivilegedActionException |
-                         InvocationTargetException |
-                         IllegalAccessException e) {
-                    assert !checkpointException.hasException() :
-                        "should not have new arguments";
-                    restoreException.handle(e);
-                }
+        if (newArguments != null && newArguments.length > 0) {
+            try {
+                Method newMain = AccessController.doPrivileged(new PrivilegedExceptionAction<Method>() {
```
[the remark this answers is not in the document]
Comment (YizhePKU) [reply src/java.base/share/classes/jdk/crac/Core.java:233]: It's a good idea, but I think it's better to leave this to another PR.

File: src/hotspot/share/runtime/arguments.cpp [hunk 4]
```diff
@@ -2283,7 +2283,11 @@ bool Arguments::parse_options_for_restore(const JavaVMInitArgs* args) {
         if (old_java_command != NULL) {
           os::free(old_java_command);
         }
-      } else {
+      }
+      else if (strcmp(key, "CRaCJavaMainArgs") == 0) {
```
[the remark this answers is not in the document]
Comment (YizhePKU) [reply src/hotspot/share/runtime/arguments.cpp]: Done.

**Comment by YizhePKU** [comment]
I've yet to get existing tests to pass on my machine (it's NixOS, so it never works the first time). I'll add the tests after I figure it out.

File: src/hotspot/share/runtime/arguments.cpp [hunk 5]
```diff
@@ -2283,7 +2283,11 @@ bool Arguments::parse_options_for_restore(const JavaVMInitArgs* args) {
         if (old_java_command != NULL) {
           os::free(old_java_command);
         }
-      } else {
+      }
+      else if (strcmp(key, "jdk.internal.crac.mainArgs") == 0) {
```
Comment (AntonKozlov) [inline src/hotspot/share/runtime/arguments.cpp:2287]: Nit: please use One-True-Brace-Style https://github.com/openjdk/jdk/blob/master/doc/hotspot-style.md#whitespace

**Review by AntonKozlov — COMMENTED** [review]
Would you mind rebasing the PR?
